modules/@apostrophecms/doc-type/index.js

```diff
@@ -2050,41 +2051,61 @@ module.exports = {
         // the distinct values and whose values are the counts.
         // This has a performance impact. Not chainable.
 
-        async toDistinct(property) {
-          await query.finalize();
+        toDistinct(property) {
+          const telemetry = self.apos.telemetry;
+          return telemetry.aposStartActiveSpan(`db:${self.__meta.name}:query:toDistinct`, async (span) => {
+            span.setAttribute(SemanticAttributes.CODE_FUNCTION, 'toDistinct');
+            span.setAttribute(SemanticAttributes.CODE_NAMESPACE, self.__meta.name);
 
-          if (!query.get('distinctCounts')) {
-            return self.apos.doc.db.distinct(property, query.get('criteria'));
-          } else {
-            return distinctCounts();
-          }
-
-          async function distinctCounts() {
-            const pipeline = [
-              {
-                $match: query.get('criteria')
-              },
-              {
-                $unwind: '$' + property
-              },
-              {
-                $group: {
-                  _id: '$' + property,
-                  count: {
-                    $sum: 1
+            async function distinctCounts(span) {
+              const pipeline = [
+                {
+                  $match: query.get('criteria')
+                },
+                {
+                  $unwind: '$' + property
+                },
+                {
+                  $group: {
+                    _id: '$' + property,
+                    count: {
+                      $sum: 1
+                    }
                   }
                 }
+              ];
+              span.setAttribute(SemanticAttributes.DB_STATEMENT, JSON.stringify({ pipeline }, null, 2));
+              const results = await self.apos.doc.db.aggregate(pipeline).toArray();
+              const counts = {};
+
+              _.each(results, function(doc) {
+                counts[doc._id] = doc.count;
+              });
+              query.set('distinctCounts', counts);
+              return results.map(result => result._id);
+            }
+
+            try {
+              await query.finalize();
+              let result;
+
+              if (!query.get('distinctCounts')) {
+                const criteria = query.get('criteria');
+                span.setAttribute(SemanticAttributes.DB_STATEMENT, JSON.stringify({ criteria }, null, 2));
```

Comment:
util.stringify is more appropriate because it provides a readable representation of regular expressions.

Reply:
NodeJS `util.inspect` I assume as I'm unable to find any `apos.util.stringify` method.